pkg/kinflate/transformers/overlay.go

```diff
@@ -95,3 +106,59 @@ func (o *overlayTransformer) Transform(baseResourceMap resource.ResourceCollecti
 	}
 	return nil
 }
+
+// mergePatches merge and index patches by GVKN.
+// It errors out if there is conflict between patches.
+func (o *overlayTransformer) mergePatches() (resource.ResourceCollection, error) {
+	rc := resource.ResourceCollection{}
+	patches := resourcesToObjects(o.overlay)
+	for ix, patch := range o.overlay {
+		gvkn := patch.GVKN()
+		existing, found := rc[gvkn]
+		if !found {
+			rc[gvkn] = patch
+			continue
+		}
+
+		versionedObj, err := scheme.Scheme.New(gvkn.GVK)
+		if err != nil && !runtime.IsNotRegisteredError(err) {
+			return nil, err
+		}
+		var cd conflictDetector
+		if err != nil {
+			cd = newJMPConflictDetector()
+		} else {
+			cd, err = newSMPConflictDetector(versionedObj)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		conflict, err := cd.hasConflict(existing.Data, patch.Data)
+		if err != nil {
+			return nil, err
+		}
+		if conflict {
+			conflictingPatch, err := cd.findConflict(ix, patches)
```

Comment:
Suppose you have patches A, B, C and D. And A conflicts with D.
After merging A, B and C, you will have (A+B+C) which is `existing.Data`.
Then you find merged patch (A+B+C)  conflicts with D.
You need to use the patches before merging to find out A conflicts with D. So you need to pass in the conflicting patch (it can be identified by `ix`) and all of the patches (`patches`).
IOW, we need n patches instead of 2. `existing.Data` and `patch.Data` are 2 patches, we can not get some useful error from it.

Reply:
The explanation helps. Thank you